chore: switch to ruff for python projects

sfoster1 opened the pull request:
Ruff ( https://docs.astral.sh/ruff ) is a modern code checking tool that combines many formatting and style/correctness linting tools and is really fast. It has a lot of options for things that have previously been flake8 plugins or separate tools, and it's configured in pyproject.toml instead of a separate file.

It also has a lot of rules ( https://docs.astral.sh/ruff/rules )
for things we've wanted forever but haven't found a good tool for yet - like import sorting!

Its formatting codestyle philosophy is very similar to black - not a lot of options, which is good - and so is its formatting output... but its formatting output is not _identical_ to black's ( https://docs.astral.sh/ruff/formatter/black/ ), and so there is a big old diff here that is 99% automatic and 1% "that type: ignore or noqa is in the wrong place now".

The PR is split into commits and probably should be read that way - in fact, I don't know how much you're going to get out of reading the subsequent commits that apply the formatting, but be your own judge.

The first commit is the key; it contains the addition of ruff and the porting of the flake8 rules to the ruff config in pyproject.toml.

Future work includes enabling import sorting and poking around other kinds of linting that is available in ruff but not yet used - this just added the equivalent of our black + flake8 + flake8 plugin testing.

sfoster1 committed:
chore: switch mainline python projects to ruff

Ruff ( https://docs.astral.sh/ruff ) is a modern code checking tool that
combines many formatting and style/correctness linting tools and is
really fast. It has a lot of options for things that have previously
been flake8 plugins or separate tools, and it's configured in
pyproject.toml instead of a separate file.

It also has a lot of options for things we've wanted forever but haven't
found a good tool for yet - like import sorting!

Its formatting codestyle philosophy is very similar to black - not a lot
of options, which is good - and so is its formatting output... but its
formatting output is not _identical_ to black's, and so there is a big
old diff here that is 99% automatic and 1% "that type: ignore or noqa is
in the wrong place now".

sfoster1 committed:
chore: apply ruff to api

Giant diff, no code changes.

sfoster1 committed:
chore(g-code-testing): apply ruff

sfoster1 committed:
chore(hardware): apply ruff

sfoster1 committed:
chore(robot-server): Apply ruff

sfoster1 committed:
chore(server-utils): Apply ruff

sfoster1 committed:
chore(shared-data): apply ruff

sfoster1 committed:
chore(system-server): apply ruff

sfoster1 committed:
chore(update-server): apply ruff

sfoster1 committed:
chore(usb-bridge): apply ruff

ddcc4 reviewed (approved):
I spot-checked some of the files in `api/`.

Your new formatter seems to tolerate slightly wider lines than the old one? which I think is causing a the re-wrapping.

sfoster1 commented:
> I spot-checked some of the files in `api/`.
> 
> Your new formatter seems to tolerate slightly wider lines than the old one? which I think is causing a the re-wrapping.

probably. i'm a big believer in the black philosophy of "one big diff is better than a lifetime of nitpicking format rules" so I want to roll with it.

sfoster1 commented:
Somehow a snapshot test didn't get updated for py 3.12 yet

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal switch-to-ruff snapshots (#20452)

This PR was requested on the PR
https://github.com/Opentrons/opentrons/pull/20451

Co-authored-by: sfoster1 <3091648+sfoster1@users.noreply.github.com>
